### PR TITLE
Add script to copy a resource and convert core questions to library interactives.

### PR DESCRIPTION
**Pull request by emcelroy**
[#174824504]

### CODE REVIEW

**Comment by emcelroy** [comment]
@kswenson Is what I've done in moving the types into separate files pretty much what you were thinking? Things got a little tricky with the drawing tool and image question. I'm not confident I did all that correctly.

**Review by kswenson — CHANGES_REQUESTED** [review]
>Is what I've done in moving the types into separate files pretty much what you were thinking? Things got a little tricky with the drawing tool and image question. I'm not confident I did all that correctly.

Basically, what you did with moving types into separate files is exactly what I had in mind. There are a couple of issues with the implementation, however.

The main one is that @pjanik made a bunch of changes to the drawing tool and image question types in #61, #62, and #63. When we talked about this I didn't realize how old this branch was and that it doesn't include any of these recent PRs. I think the best approach at this point would be to back out of your type-related changes to those interactives, merge from master, and then take another crack at it.

The other issue is that in several cases you're redefining types that were already defined in the base type. In cases where the types are the same the redefinition is simply redundant. In cases where the types are different, it's important to understand why and whether the proposed type changes are appropriate. If the types are being redefined because they're wrong in the base type, then they should be fixed in the base type in the library instead. On the other hand, redefining the `question_type` to `any` seems like an inappropriate widening of the type -- it doesn't make sense to pass `any` value for `question_type`.

The upshot is that ideally, the conversion script wouldn't require any changes to the preexisting types. Where type changes are required. e.g. because the conversion script revealed missing properties, those changes should be made in the appropriate place, which may be the `interactive-api` library in some cases. In other cases, the right solution may be to use type casts in the conversion script itself, for instance, rather than widening the types in the interactive definitions. Ultimately, any such proposed type changes should be considered on the merits.

File: src/image/components/types.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// Note that TS interfaces should match JSON schema. Currently there's no way to generate one from the other.
+// TS interfaces are not available in runtime in contrast to JSON schema.
+
+export interface IAuthoredState {
+  version: number;
+  questionType: any;
```
[the remark this answers is not in the document]
Comment (emcelroy) [reply src/image/components/types.ts]: If I remove `questionType` from here, I get this error:

`Object literal may only specify known properties, and 'questionType' does not exist in type 'IAuthoredState'.`
  
There isn't an `IAuthoringImageMetadata` in interactive-api. If I import `IAuthoringInteractiveMetadata` into components/image/types.ts, I get this error:

`Type '"image"' is not assignable to type '"iframe_interactive"'.`

Would it be best to add an `IAuthoringImageMetadata` to interactive-api that defines `questionType`?

File: src/multiple-choice/components/types.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+export interface IAuthoredState extends IAuthoringMultipleChoiceMetadata {
+  version: number;
+  questionType: any;
+  hint?: string;
+  multipleAnswers?: boolean;
+  layout?: ILayout;
+  enableCheckAnswer?: boolean;
+  customFeedback?: boolean;
+  choices: IChoice[];
```
[the remark this answers is not in the document]
Comment (emcelroy) [reply src/multiple-choice/components/types.ts:22]: If I remove `choices: IChoice[];` from here, I get this error:

`Property 'choiceFeedback' does not exist on type 'IAuthoringMultipleChoiceChoiceMetadata'.`

I don't know if it makes sense for `choiceFeedback` to be defined here or the `interactive-api` library. @scytacki Do you have any thoughts on that?

**Comment by emcelroy** [comment]
@kswenson Thanks for the thorough review and explanation. I believe I've done most of what you suggested. I had just a couple questions that I've left inline with two of your comments.

File: src/multiple-choice/components/types.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kswenson) [reply src/multiple-choice/components/types.ts:21]: I'd defer to @pjanik on this one.

File: src/image/components/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kswenson) [reply src/image/components/types.ts]: Ultimately, I would think that having the image interactive follow the pattern of the other interactives makes the most sense, i.e. have the `interactive-api` define `IAuthoringImageInteractiveMetadata` which defines `question-type: "image"`. In the short term, however, I think you can import `IAuthoringMetadataBase` and extend it to define it that way locally, which is better than widening it to `any`. Something like:
```typescript
export interface IAuthoredState extends IAuthoringMetadataBase {
  version: number;
  questionType: "image";
  ...
```

[the remark this answers is not in the document]
Comment (pjanik) [reply src/image/components/types.ts]: It's already defined in interactive-api:
https://github.com/concord-consortium/lara/blob/master/lara-typescript/src/interactive-api-client/metadata-types.ts#L42-L45
The questionType is "image_interactive", not just "image".

File: src/multiple-choice/components/types.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pjanik) [reply src/multiple-choice/components/types.ts:21]: interactive-api defines only values that are necessary for reporting. I guess choiceFeedback is not, it's specific to this interactive, so it should be defined here.

File: src/image/components/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (emcelroy) [reply src/image/components/types.ts]: Thanks, @pjanik. Is that right, though? There is an image-question interactive, but also a separate image interactive. Should the two share `IAuthoringImageQuestionMetadata`?

File: src/open-response/components/types.ts [hunk 3]
```diff
@@ -7,6 +7,8 @@ export interface IAuthoredState extends IAuthoringOpenResponseMetadata {
   version: number;
   defaultAnswer?: string;
   hint?: string;
+  predictionFeedback?: string;
+  customFeedback?: string;
```
Comment (pjanik) [inline src/open-response/components/types.ts:11]: Is `customFeedback` necessary?

File: src/image/components/runtime.test.tsx [hunk 4]
```diff
@@ -18,6 +19,7 @@ const authoredState: IAuthoredState = {
 };
 const naturalWidthImageAuthoredState: IAuthoredState = {
   version: 1,
+  questionType: "image",
```
Comment (pjanik) [inline src/image/components/runtime.test.tsx:22]: Image interactive is not a question, so it doesn't need that.

File: src/image/components/types.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pjanik) [reply src/image/components/types.ts]: Right, I confused this with the image **question** interactive.
Regular image interactive (not question) shouldn't use any metadata interface and it shouldn't define questionType.

@kswenson,
> Ultimately, I would think that having the image interactive follow the pattern of the other interactives makes the most sense, i.e. have the interactive-api define IAuthoringImageInteractiveMetadata which defines question-type: "image".
Image non-question interactive isn't a question, so it shouldn't really use any metadata interface.

Metadata interfaces are only meant for interactives that pretend to be a specific kind of question and want to be reported in a special way (without loading the whole iframe in reporting mode). Image interactive isn't a question, so it doesn't really need to use any metadata interface and it doesn't have to specify `questionType` at all.

File: src/open-response/components/types.ts (hunk 3 above)
Comment (emcelroy) [reply src/open-response/components/types.ts]: Thanks, @pjanik. You're exactly right about how the new OR interactive simplifies things. I tested it out and it worked as predicted.

I went ahead with your suggestion for setting `predictionFeedback` in the conversion script.

**Comment by scytacki** [comment]
@kswenson can you re-review this soon?

**Review by scytacki — APPROVED** [review]
Seems good to me. I haven't looked at it in detail, but from the comments I see others have.

**Review by kswenson — APPROVED** [review]
👍 Looks good

>@kswenson can you re-review this soon?

Sorry, I wasn't aware that further review was required.

**Comment by scytacki** [comment]
@pjanik did Ethan address all of your comments?

File: src/convert-old-lara/convert.ts [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+  const authoredState: IImageQuestionAuthoredState = {
+    version: 1,
+    questionType: "image_question",
+    required: item.embeddable.is_prediction,
+    predictionFeedback: item.embeddable.prediction_feedback,
+    backgroundImageUrl: item.embeddable.bg_url,
+    imageFit: "shrinkBackgroundToCanvas",
+    imagePosition: "center",
+    answerPrompt: item.embeddable.prompt,
+    prompt: item.embeddable.drawing_prompt,
```
Comment (pjanik) [inline src/convert-old-lara/convert.ts:91]: @emcelroy, I think `answerPrompt` and `prompt` should be assigned in the opposite way:
```
 answerPrompt: item.embeddable.drawing_prompt,
 prompt: item.embeddable.prompt,
```
I'm 90% sure, so you could double check in LARA.

File: src/convert-old-lara/convert.ts [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+    hint: item.embeddable.hint,
+    customFeedback: item.embeddable.give_prediction_feedback
```
Comment (pjanik) [inline src/convert-old-lara/convert.ts:93]: I think customFeedback is not necessary.

File: src/image-question/components/types.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  IAuthoringImageQuestionMetadata, 
+  IRuntimeImageQuestionMetadata 
+} from "@concord-consortium/lara-interactive-api";
+
+export interface IAuthoredStateBase extends IAuthoringImageQuestionMetadata {
+  // IRuntimeImageQuestionMetadata adds:
+  // - answerType: "image_question"
+  // - answerPrompt?: string;
+  defaultAnswer?: string;
+  customFeedback?: string;
```
Comment (pjanik) [inline src/image-question/components/types.ts:16]: I don't think customFeedback is used anywhere in image question.

**Review by pjanik — CHANGES_REQUESTED** [review]
I think my old comments are resolved now (I'll mark them so). 
I've looked more at convert.ts this time and found there a few possible issues. All described in inline comments. I focused on properties conversion and skipped the whole mechanics of this page, I'm assuming that if it works, it works. :)

File: src/convert-old-lara/convert.ts (hunk 5 above)
Comment (emcelroy) [reply src/convert-old-lara/convert.ts]: You're right. I kept getting those confused. That's fixed now.

**Review by pjanik — APPROVED** [review]
Looks good, thanks!

**Pull request merged**